### DISCUSSION TITLE
There are no commands defined in the “press” namespace

### DISCUSSION THREAD

**Discussion by skfahad** [laravel/framework#36797] [Q&A]
I am creating my own laravel package. In that package I have to register a custom command which does nothing but print a text in the console.

[By following this guide][1] I have created my command file inside `src\Console` folder and registered this command in 'boot' method of my `ServiceProivider`.

I have installed my package into my laravel project through symlink and it is being shown in the vendor directory of my laravel project.

When I run `php artisan` in laravel project, my custom command doesn't appear in the command list. When I execute command as `php artisan press:process` I get the following error

    There are no commands defined in the "press" namespace.

**Code from laravel package**

*composer.json*

    "autoload": {
        "psr-4": {
            "fahad\\Press\\": "src/"
        }
    }

*ProcessCommand.php*

    <?php
    
    namespace fahad\Press\Console;
    
    use Illuminate\Console\Command;
    
    class ProcessCommand extends Command
    {
        protected $signature = 'press:process';
    
        protected $description = 'Updates blog post';
    
        public function handle()
        {
            $this->info('Hello');
        }
    }

*PressBaseServiceProvider.php*

    <?php
    
    namespace fahad\Press;
    
    use fahad\Press\Console\ProcessCommand;
    use Illuminate\Support\ServiceProvider;
    
    class PressBaseServiceProvider extends ServiceProvider
    {
        public function boot()
        {
            $this->registerResources();

            if ($this->app->runningInConsole()) {
                $this->commands([
                    ProcessCommand::class
                ]);
            }
        }
    }

Since I'm using **Laravel 8** I have even tried to put my command file into `src\Console\Commands` to match up with laravel directory structure but it didn't work.

  [1]: https://laravelpackage.com/06-artisan-commands.html#artisan-commands

**Comment by selcukcukur** [comment]
Add the function `registerCommand()`, `provides()` and `register()` into `PressBaseServiceProvider.php`. 

```
public function register()
{
       $this->registerCommand();
}

// Let's save the command in the application container.
protected function registerCommand()
{
     $this->app->singleton('command.press.process', function($app) {

         return new ProcessCommand();
     });
}

// Let's declare the alias of the command to the application through the service provider.
public function provides()
{
      return [
          'command.press.process'
      ];
}
```

Note that if you are creating an artisan command it must be saved in the application's container. Otherwise, the command will not work correctly.

**Reply by skfahad** [reply]
I am still getting the same error

**Comment by selcukcukur** [comment]
Although the `PressBaseServiceProvider.php` file is as follows?
```
<?php

namespace fahad\Press;

use fahad\Press\Console\ProcessCommand;
use Illuminate\Support\ServiceProvider;

class PressBaseServiceProvider extends ServiceProvider
{
    public function boot()
    {
        $this->registerResources();

        if ($this->app->runningInConsole()) {
            $this->commands([
                ProcessCommand::class
            ]);
        }
    }

    public function register()
    {
          $this->registerCommand();
    }

    // Let's save the command in the application container.
   protected function registerCommand()
   {
       $this->app->singleton('command.press.process', function($app) {

         return new ProcessCommand();
       });
   }

   // Let's declare the alias of the command to the application through the service provider.
   public function provides()
   {
        return [
            'command.press.process'
        ];
   }
}
```

Are you adding the service provider to the configuration file, right?

**Reply by skfahad** [reply]
```
"extra": {
        "laravel": {
            "dont-discover": [],
            "providers": [
                "fahad\\Press\\PressBaseServiceProvider"
            ]
        }
    },
```

I have added the above lines of code into my laravel project's composer.json file which is responsible for auto discovering my service provider

**Reply by selcukcukur** [reply]
Would you try and update `ProcessCommand.php` like this?
```
<?php

namespace fahad\Press\Console;

use Illuminate\Console\Command;

class ProcessCommand extends Command
{
    protected $signature = 'press:process';

    protected $description = 'Updates blog post';

    public function __construct()
    {
        parent::__construct();
    }

    public function handle()
    {
        $this->info('Hello');
    }
}
```

**Reply by skfahad** [reply]
Actually I got the solution. It was a silly mistake, I added the extra part in app's composer instead of package's composer and that's why my service provider was never being called
